generics-sop/src/Generics/SOP/Universe.hs

```diff
@@ -185,6 +185,47 @@ productTypeTo :: IsProductType a xs => NP I xs -> a
 productTypeTo = to . SOP . Z
 {-# INLINE productTypeTo #-}
 
+-- | Constraint that captures that a datatype is a (simple) sum type,
+-- i.e., a type with some number of constructors, each of which 
+-- has a single argument.
+--
+-- It also gives access to the list of types which make up the union.
+--
+-- @since 0.5.2.0
+--
+type IsSumType (a :: Type) (xs :: [Type]) =
+  (Generic a, AllZip IsSingletonOf xs (Code a))
+
+-- | Direct access to the list of types that makes up a sum type.
+--
+-- @since 0.5.2.0
+--
+type SumCode (a :: Type) = Heads (Code a)
+
+-- | Convert from a sum type to its sum representation.
+--
+-- @since 0.5.2.0
+--
+sumTypeTo :: IsSumType a xs => a -> NS I xs
+sumTypeTo = go . unSOP . from
+  where
+    go :: AllZip IsSingletonOf xs xss => NS (NP I) xss -> NS I xs
+    go (Z (x :* Nil)) = Z x
+    go (S xss) = S $ go xss
+{-# INLINE sumTypeTo #-}
+
+-- | Convert a sum representation to the original type.
+--
+-- @since 0.5.2.0
+--
+sumTypeFrom :: IsSumType a xs => NS I xs -> a
+sumTypeFrom = to . SOP . go
+  where
+    go :: AllZip IsSingletonOf xs xss => NS I xs -> NS (NP I) xss
+    go (Z x) = Z (x :* Nil)
+    go (S xss) = S $ go xss
```

Comment:
I'd much rather have the conversion functions be applications of `trans_NS`.

This is simple enough for `sumTypeFrom`:
```haskell
sumTypeFrom :: IsSumType a xs => a -> NS I xs
sumTypeFrom = to . SOP . trans_NS (Proxy :: Proxy IsSingletonOf) (:* Nil)
```

However, we need the flipped constraint for `sumTypeTo`:
```haskell
sumTypeTo = trans_NS (Proxy :: Proxy (Flip IsSingletonOf)) hd . unSOP . from
```

This requires the obvious class/instance definition of `Flip`, but also means we have to add `AllZip (Flip IsSingletonOf) (Code a) xs` to the constraints for `IsSumType`.

This is a general annoyance with `trans_NP` and `AllZip`, the GHC cannot prove one constraint from the other. There are different options here:

- add a version of `AllZip` that expands to both directions,
- add a variant `rtrans_*` to the `trans_*` family of functions which works on the reversed constraint,
- use explicit dictionary manipulation to turn one constraint into the other.

I'm not sure what's best.

Reply:
I haven't forgotten about this, but it'll take me a while to try either of those, and I certainly don't feel like I have the judgement to say which of your suggested alternatives is better.